File: 프로그래머스/알고리즘 고득점 Kit/스택, 큐/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck):
    bridge = deque([0]*bridge_length)
    truck = deque(truck)
    total_weight = 0
    time = 0
    
    while truck or total_weight:
        total_weight -= bridge.popleft()

        if truck and total_weight + truck[0] <= weight:    
            next = truck.popleft()
            total_weight += next
            bridge.append(next)
        else:
            bridge.append(0)
        
        time += 1

    return time

File: 프로그래머스/알고리즘 고득점 Kit/스택, 큐/다리를 지나는 트럭.js
class Deque {
	constructor() {
		this.storage = {};
		this.front = 0;
		this.rear = 0;
	}

	size() {
		return this.rear - this.front;
	}

	append(item) {
		this.storage[this.rear++] = item;
	}

	pop() {
		if (!this.size()) return null;
		const item = this.storage[--this.rear];
		delete this.storage[this.rear];
		return item;
	}

	popleft() {
		if (!this.size()) return null;
		const item = this.storage[this.front];
		delete this.storage[this.front++];
		return item;
	}
}

function solution(bridge_length, weight, truck) {
	const bridge = new Deque();
	for (let i = 0; i < bridge_length; i++) bridge.append(0);
	const trucks = new Deque();
	for (const truck_weight of truck) trucks.append(truck_weight);

	let total_weight = 0;
	let time = 0;

	while (trucks.size() || total_weight) {
		total_weight -= bridge.popleft();

		if (trucks.size() && total_weight + trucks.storage[trucks.front] <= weight) {
			let incoming_truck = trucks.popleft();
			total_weight += incoming_truck;
			bridge.append(incoming_truck);
		} else {
			bridge.append(0);
		}

		time += 1;
	}

	return time;
}
